scripts/yyParticle.js

```diff
@@ -236,6 +242,150 @@ function ParticleSystem_ClearClass()
+	this.emitters = [];
+}
+
+/// All existing particle systems.
+CParticleSystem.instances = [];
+
+/// <returns>A new particle system.</returns>
+CParticleSystem.Create = function ()
+{
+	var system = new CParticleSystem();
+	system.index = CParticleSystem.GetCount();
+	CParticleSystem.instances.push(system);
+	return system;
+};
+
+/// <summary>
+/// Creates a new particle system, loading its data from JSON.
+/// </summary>
+/// <param name="json">A particle system entry from the JSON.</param>
+/// <returns>The created particle system.</returns>
+CParticleSystem.CreateFromJSON = function (json)
+{
+	var system = CParticleSystem.Create();
+	system.name = json.pName;
+	system.originX = json.originX;
+	system.originY = json.originY;
+	system.drawOrder = json.drawOrder;
+	for (var i = 0; i < json.emitters.length; ++i)
+	{
+		system.emitters.push(json.emitters[i]);
+	}
+	return system;
+};
+
+/// <returns>Total number of existing particle system resources.</returns>
+CParticleSystem.GetCount = function ()
+{
+	return CParticleSystem.instances.length;
+};
+
+/// <summary>
+/// Retrieves a particle system with given index.
+/// </summary>
+/// <param name="index">The index of the particle system.</param>
+/// <returns>The particle system resource or NULL if it doesn't exist.</returns>
+CParticleSystem.Get = function (index)
+{
+	if (index < 0 || index >= CParticleSystem.GetCount())
+	{
+		return null;
+	}
+	return CParticleSystem.instances[index];
+};
+
+/// <summary>
+/// Retrieves an index of a particle system with given name.
+/// </summary>
+/// <param name="name">The name of the particle system.</param>
+/// <returns>The index or -1 on fail.</returns>
+CParticleSystem.Find = function (name)
+{
+	for (var i = CParticleSystem.GetCount() - 1; i >= 0; --i)
+	{
+		if (name == CParticleSystem.instances[i].name)
+		{
+			return i;
+		}
+	}
+	return -1;
+};
+
+/// <returns>The index of the particle system.</returns>
+CParticleSystem.prototype.GetIndex = function ()
+{
+	return this.index;
+};
+
+
+/// <summary>
+/// Creates an instance of the particle system.
+/// </summary>
+/// <param name="_layerID"></param>
+/// <param name="_pParticleEl">The layer to use. A new one is created if not defined.</param>
+/// <param name="_persistent"></param>
+/// <returns>The index of the created instance.</returns>
+CParticleSystem.prototype.MakeInstance = function (_layerID, _persistent, _pParticleEl)
+{
+	if (_layerID === undefined) _layerID = -1;
+	if (_persistent === undefined) _persistent = false;
+	if (_pParticleEl === undefined) _persistent = null;
+
+	var ps = (_pParticleEl == null)
+		? ParticleSystem_Create(_layerID, _persistent)
+		: ParticleSystem_Create_OnLayer(-1, _persistent, _pParticleEl);
+	
+	var system = g_ParticleSystems[ps];
+
+	for (var i = 0; i < this.emitters.length; ++i)
+	{
+		var emitterIndex = this.emitters[i];
+
+		var templateEmitter = g_PSEmitters[emitterIndex];
+
+		var em = ParticleSystem_Emitter_Create(ps);

```

Comment:
Should we be checking for failure here? If it's not found and returns -1?

Reply:
Good catch 👍